impl/src/main/java/io/kroxylicious/testing/kafka/common/ListeningSocketPreallocator.java

```diff
@@ -7,39 +7,88 @@
 
 import java.io.IOException;
 import java.io.UncheckedIOException;
+import java.net.BindException;
 import java.net.ServerSocket;
 import java.util.ArrayList;
 import java.util.List;
-import java.util.stream.Stream;
+import java.util.Objects;
+import java.util.Random;
+import java.util.stream.Collectors;
 
 /**
  * Allocates one or more groups listening sockets from the ephemeral port range.
- *
+ * <p>
  * The caller may call #preAllocateListeningSockets any number of times.
  * The caller is guaranteed that all allocated port numbers are unique, even across invocations of #preAllocateListeningSockets.
- *
+ * <p>
  * Before the caller attempts to bind, they must call #close.  Once close is called no further use of the socket
  * allocator is permitted.
  */
 public class ListeningSocketPreallocator implements AutoCloseable {
 
+    private static final int PORT_RANGE_LOW = 10_000;
+    // Ideally this would be derived from `sysctl net.inet.ip.portrange.first` or `sysctl net.ipv4.ip_local_port_range` depending on the platform
+    // however that's not so easily achieved, so we hard code instead.
+    private static final int PORT_RANGE_HIGH = 30_000;
+    private static final int MAX_PORT_COUNT = PORT_RANGE_HIGH - PORT_RANGE_LOW;
+
     private final List<ServerSocket> all = new ArrayList<>();
 
+    private final Random random = new Random();
+
     /**
      * Instantiates a new Listening socket preallocator.
      */
     public ListeningSocketPreallocator() {
     }
 
+    /**
+     * Pre-allocate 1 or more ports from a defined range, to avoid collisions without going network connections,
+     * which are available for use once the #close method is called.
+     *
+     * @param num number of ports to pre-allocate
+     * @return a List of bound sockets
+     */
+    public List<ServerSocket> preAllocateListeningSockets(int num) {
```

Comment:
Not convinced by returning a list here as the allocator doesn't care about the order, but every invocation of the stream version turned it into a list (as they use index based access) so it seemed like the expedient choice here.